airbyte-integrations/bases/base-normalization/normalization/transform_catalog/catalog_processor.py

```diff
@@ -0,0 +1,183 @@
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+"""
+
+import json
+import os
+from typing import Dict, Set
+
+import yaml
+from normalization.destination_type import DestinationType
+from normalization.transform_catalog.destination_name_transformer import DestinationNameTransformer
+from normalization.transform_catalog.stream_processor import StreamProcessor
+from normalization.transform_catalog.utils import jinja_call
+
+
+class CatalogProcessor:
+    """
+    Takes as input an AirbyteCatalog file (stored as Json Schema).
+    Associated input raw data is expected to be stored in a staging area called "raw_schema".
+
+    This processor reads the catalog file, extracts streams descriptions and transforms them to final tables in their
+    targeted destination schema.
+
+    This is relying on a StreamProcessor to handle the conversion of a stream to a table one at a time.
+    """
+
+    def __init__(self, output_directory: str, integration_type: DestinationType):
+        """
+        @param output_directory is the path to the directory where this processor should write the resulting SQL files (DBT models)
+        @param integration_type is the destination type of warehouse
+        """
+        self.output_directory: str = output_directory
+        self.integration_type: DestinationType = integration_type
+        self.name_transformer: DestinationNameTransformer = DestinationNameTransformer(integration_type)
+
+    def process(self, catalog_file: str, json_column_name: str, target_schema: str):
+        """
+        This method first build the top-level streams and in a second loop will go over the substreams that
+        were nested. (breadth-first traversal)
+
+        @param catalog_file input AirbyteCatalog file in JSON Schema describing the structure of the raw data
+        @param json_column_name is the column name containing the JSON Blob with the raw data
+        @param target_schema is the final schema where to output the final transformed data to
+        """
+        # Registry of all tables in all schemas
+        tables_registry: Set[str] = set()
+        # Registry of source tables in each schemas
+        source_tables: Dict[str, Set[str]] = {}
+
+        catalog = read_json(catalog_file)
+        substreams = {}
+        for configured_stream in get_field(catalog, "streams", "Invalid Catalog: 'streams' is not defined in Catalog"):
+            stream_config = get_field(configured_stream, "stream", "Invalid Stream: 'stream' is not defined in Catalog streams")
+            schema_name = self.name_transformer.normalize_schema_name(target_schema)
+            raw_schema_name = self.name_transformer.normalize_schema_name(f"_airbyte_{target_schema}")
```

Comment:
Actually, this new schema is not created and left empty if the DBT project settings is put to "ephemeral"

If you set it to "view", then it will be populated.

I'll make a video explaining this

Reply:
okay. my understanding is that these ephemeral tables are put in a separate schema. in practice this doesn't change anything for the end user because they are never materialized. but if we are debugging we can materialize them if we want to. the final tables are still put in their original schemas.  assuming that what i am saying is correct then i think we are all good here.